sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -610,18 +610,13 @@ trait HiveTypeCoercion {
    */
   object Division extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
-      // Skip nodes who's children have not been resolved yet.
-      case e if !e.childrenResolved => e
+      // Skip nodes who's children have not been resolved yet or input types do not match.
+      case e if !e.childrenResolved || e.checkInputDataTypes().hasError => e
```

Comment:
as discussed offline, maybe it'd make sense to have resolved also call the data check function.